### PR TITLE
feat: support `system.parquet_files` table

### LINKED ISSUE #24988: Create the `system.parquet_files` provider
Add a new system table for storing information about parquet files stored by `influxdb3`. This table should only provide files related to a specified database name and table name, to limit the scope of data returned.

The columns to include in this table can be based on what we store in segment files, e.g.,
* `path` (object store file path)
* `size_bytes`
* `row_count`
* `min_time`
* `max_time`

### PR DESCRIPTION
**Pull request by hiltontj**
This PR is intended to follow #25000.

Extend system table support to allow queries against `system.parquet_files`, a new system table that allows queries for parquet files associated with a given database and table.

Currently, the query will only look back on 1,000 segments to get associated parquet files, we may want to adjust that, or make it more flexible.

Two unit tests were added to `query_executor` for this, instead of E2E integration tests, as more control was needed over file persistence than what the E2E test harness provides.

## Example
The `system.parquet_files` table can be queried like so:
```
SELECT * FROM system.parquet_files WHERE table_name = 'cpu'
```
And responds accordingly:
```
+------------+-------------------------------------------------+------------+-----------+---------------------+---------------------+
| table_name | path                                            | size_bytes | row_count | min_time            | max_time            |
+------------+-------------------------------------------------+------------+-----------+---------------------+---------------------+
| cpu        | dbs/foo/cpu/2024-05-14T20-53/4294967294.parquet | 1775       | 18        | 1715720025774406000 | 1715720035952490000 |
+------------+-------------------------------------------------+------------+-----------+---------------------+---------------------+
```

## Other Changes

Another change in this PR was to go back to using dynamic dispatch for passing around the `Persister`. The associated types and generics made using the `Persister` a pain in the `ParquetFilesTable` implementation, so I removed those in favour of using `Arc<dyn Persister>`, as we had been doing previously (see https://github.com/influxdata/influxdb/pull/24705/commits/42d36d575b16011be5bdb0b9250e2b818d3a59b5).

Closes #24988

### CODE REVIEW

File: influxdb3_write/src/lib.rs [hunk 1]
```diff
@@ -72,6 +72,8 @@ pub trait WriteBuffer: Bufferer + ChunkContainer {}
 /// the WAL is configured. As a segment is closed, it is persisted and then freed from memory.
 #[async_trait]
 pub trait Bufferer: Debug + Send + Sync + 'static {
+    type PersisterError;
```
Comment (hiltontj) [inline influxdb3_write/src/lib.rs:75]: This should not be added.

**Review by pauldix — COMMENTED** [review]
The SegmentState in the write buffer actually has the information for the last 1,000 persisted segments and their parquet files: https://github.com/influxdata/influxdb/blob/main/influxdb3_write/src/write_buffer/segment_state.rs#L45. This is populated on startup.

It may be a good idea to pull the parquet files information from there, rather than the persister/object store. When the buffer gets the ability to persist some table data ahead of the segment rolling over, that state will be kept there in memory and won't show up in a segment info file until after the segment rolls over.

I think I'll be refactoring the structure of SegmentState when doing that. So maybe wait until after that to change this?

**Comment by hiltontj** [comment]
> It may be a good idea to pull the parquet files information from [the `SegmentState`]

That makes sense and seems simpler. Giving the system table provider read access to the `SegmentState` shouldn't be too hard and reading the `PersistedSegments` off of that should be much cheaper than how I'm doing it here. In fact, with making this API public, I should be able to get them directly: https://github.com/influxdata/influxdb/blob/2381cc6f1d6c289a09840024b99fada990c0a23b/influxdb3_write/src/write_buffer/segment_state.rs#L216-L232

I can at least try that out and flag if there is an issue with doing so.

**Comment by hiltontj** [comment]
@pauldix - after taking a quick look at using the `SegmentState` directly, I found it gets a bit messy with the generics on `SegmentState`. Its (truncated) definition is:
```rust
// T -> TimeProvider
// W -> Wal
struct SegmentState<T, W> {
    /* ... */
}
```
The reason it gets messy is because I need to get access via the `Bufferer` trait to whatever I need from the write buffer. So returning a thing with generics from the `Bufferer` trait becomes a nuisance. Getting the persister was easy because I could just put the `persister` method on `Bufferer`:
```rust
trait Bufferer {
    /* ... */
    fn persister(&self) -> Arc<dyn Persister>;
}
```

If you are re-thinking the segment state, it may be helpful to have a trait, e.g.,
```rust
trait SegmentStateProvider {
    fn get_parquet_files(&self) -> Vec<ParquetFile>;

    /* Other methods you are considering for this API */
}
```
Then that would be more easily extracted via the `Bufferer` trait and passed to the `SystemTableProvider` as `Arc<dyn SegmentStateProvider>`.

I can leave this for now, since you're planning to refactor, don't want to make doing so more difficult with changes here.

**Review by mgattozzi — APPROVED** [review]
This LGTM code wise. I'll defer to @pauldix though if there's anything else we need to get into this PR, but from a code perspective it looks great.

**Comment by hiltontj** [comment]
Planning to wait on https://github.com/influxdata/influxdb/issues/25144 and any related refactoring to the `influxdb3_write` crate before revisiting this.

**Comment by hiltontj** [comment]
Closing in favour of #25225

**Pull request closed without merging**